Episode on parallel raster computations (#90)
carpentries-incubator/geospatial-python: _episodes/20-parallel-raster-computations.md

---
title: "Parallel raster computations using Dask"
teaching: TODO

> exercises: TODO

Reviewer: can we estimate 40 minutes for teaching, 20 for exercises? and then revise when this is tested?

Author: Sounds good!

questions:

> - "TODO"

Reviewer: * How can I parallelize computations on rasters with dask?
* How can I determine if parallelization improves calculation speed?
* How can I control how much to parallelize calculations?

objectives:
- "Profile the timing of your raster calculations."
- "Open raster data as a chunked array."
- "Recognize good practices in selecting proper chunk sizes."
- "Setup raster calculations that take advantage of parallelization."
keypoints:

> - "TODO"

Reviewer: Here's a start:

> * "chunks" are the unit of parallelization
> * Specifying the `chunks` argument with `rioxarray.open_rasterio()` opens a raster as a chunked array
> * the chunks tuple can be adapted to parallelize calculations across different raster dimensions
> * python's `time` module or the jupyter `%%time` magic command can be used to profile calculations

Author: Wonderful, thanks! I have included these with some slight changes and added one more.

---

Very often raster computations involve applying the same operation to different pieces of data. Think, for instance, to
the "pixel"-wise sum of two raster datasets, where the same sum operation is applied to all the matching grid-cells of
the two rasters. This class of tasks can benefit from chunking the input raster(s) into smaller pieces: operations on
different data blocks can be run in parallel using multiple computing units (e.g., multi-core CPUs), thus potentially
speeding up calculations. In addition, working on chunked data can also lead to smaller memory footprints, since one
may bypass the need to store the full dataset in memory by processing it chunk by chunk.

In this episode, we will introduce the use of Dask in the context of raster calculations. Dask is a Python library for
parallel and distributed computing that provides a framework to work with different data structures, including chunked
arrays (Dask Arrays). Dask is well integrated with (`rio`)`xarray` objects, which can use Dask arrays as underlying
data structures.

> ## More Resources on Dask
>

> > TODO: Dask and Dask Arrays, with links

Reviewer: should we just link to this here? https://docs.dask.org/en/stable/array.html

Author: Added an extra link to Xarray&Dask as well!

resolution, while the scene classification layer has a lower resolution (20 m):

~~~
import rioxarray
scl = rioxarray.open_rasterio(scl_href)
visual = rioxarray.open_rasterio(visual_href)
scl.rio.resolution(), visual.rio.resolution()
~~~
{: .language-python}

~~~
((20.0, -20.0), (10.0, -10.0))
~~~
{: .output}

In order to match the image and the mask pixels, one could load both rasters and resample the finer raster to the
coarser resolution (e.g. with `reproject_match`). Instead, here we take advantage of a feature of the cloud-optimized
GeoTIFF (COG) format, which is used to store these raster files. COGs typically include multiple lower-resolution
versions of the original image, called "overviews", in the same file. This allows to avoid downloading high-resolution
images when only quick previews are required.

Overviews are often computed using powers of 2 as down-sampling (or zoom) factors (e.g. 2, 4, 8, 16). For the true-color
image we thus open the first level overview (zoom factor 2) and check that the resolution is now also 20 m:

~~~
visual = rioxarray.open_rasterio(visual_href, overview_level=0)
visual.rio.resolution()
~~~
{: .language-python}

~~~
(20.0, -20.0)
~~~

> {: .output}

Reviewer: this is excellent, valuable stuff!

Xarray and Dask also provide a graphical representation of the raster data array and of its blocked structure.

<img src="../fig/20-Dask-arrays-s2-blue-band.png" title="Xarray representation of a Dask-backed DataArray" alt="DataArray with Dask" width="612" style="display: block; margin: auto;" />

> ## Exercise: Chunk sizes matter
> We have already seen how COGs are regular GeoTIFF files with a special internal structure. Another feature of COGs is
> that data is organized in "blocks" that can be accessed remotely via independent HTTP requests, enabling partial file
> readings. This is useful if you want to access only a portion of your raster file, but it also allows for efficient
> parallel reading. You can check the blocksize employed in a COG file with the following code snippet:
>
> ~~~
> import rasterio
> with rasterio.open(cog_uri) as r:
>     if r.is_tiled:
>         print(f"Chunk size: {r.block_shapes}")
> ~~~
> {: .language-python}
>
> In order to optimally access COGs it is best to align the blocksize of the file with the chunks employed when loading
> the file. Open the blue-band asset ("B02") of a Sentinel-2 scene as a chunked `DataArray` object using a suitable

> > chunk size. Which elements do you think should be considered when choosing such value?

Reviewer: "choosing such value" > "choosing the chunk size"

>
> > ## Solution
> > ~~~
> > import rasterio
> > with rasterio.open(blue_band_href) as r:
> >     if r.is_tiled:
> >         print(f"Chunk size: {r.block_shapes}")
> > ~~~
> > {: .language-python}
> >
> > ~~~
> > Chunk size: [(1024, 1024)]
> > ~~~
> > {: .output}
> >

> > > Ideal values are thus multiples of 1024. An element to consider is the number of resulting chunks and their size.

Reviewer: this is ideal only for this dataset right? for other datasets the block size could differ I think.

"Ideal values are thus multiples of 1024" > "Ideal chunk size values for this raster are thus multiples of 1024"

Author: Yes, good point!

> > > Chunks should not be too big nor too small (i.e. too many). Recommended chunk sizes are of the order of 100 MB.

Reviewer: where does this recommendation come from? 

> Recommended chunk sizes are of the order of 100 MB.

Is this because of the typical size of a raster file and typical block size? Or is this a Dask recommendation that applies more generally to Dask arrays?

Author: For what I know, it was a rule of thumb for Dask arrays in generals. Maybe "recommendation" is too strong, I have soften it a bit by saying that this is typically a good value (and linked to the following Dask blog post: https://blog.dask.org/2021/11/02/choosing-dask-chunk-sizes). How do you see it now?

> > Also, the shape might be relevant, depending on the application! Here, we might select a chunks shape of
> > `(1, 6144, 6144)`:
> >
> > ~~~
> > band = rioxarray.open_rasterio(band_url, chunks=(1, 6144, 6144))
> > ~~~
> > {: .language-python}
> >

> > > which leads to chunks 72 MB large. Also, we can let `rioxarray` and Dask figure out appropriate chunk shapes by

Reviewer: Is my math wrong? I get 36 Mb:  (6144X6144/1024)/1024 = 36 Mb. This could be because of the numeric type of the values in the array. Can we include a python calc of the chunk size?

Author: Indeed, the number of bytes per array element is missing (the calculation is missing a "x 2", since they are `uint16`). Good point, adding the multiplication will make it much clearer..

with:

~~~
import dask
dask.visualize(visual_store)
~~~
{: .language-python}

<img src="../fig/20-Dask-arrays-graph.png" title="Dask graph" alt="dask graph" width="612" style="display: block; margin: auto;" />

The task graph gives Dask the complete "overview" of the calculation, thus enabling a better management of tasks and
resources when dispatching calculations to be run in parallel.

While most methods of `DataArray`'s run operations lazily when Dask arrays are employed, some methods by default
trigger immediate calculations, like the method `to_raster()` (we have changed this behaviour by specifying
`compute=False`). In order to trigger calculations, we can use the `.compute()` method. Again, we explicitly tell Dask
to run tasks on 4 threads. Let's time the cell execution:

~~~
%%time
visual_store.compute(scheduler="threads", num_workers=4)
~~~
{: .language-python}

~~~
CPU times: user 532 ms, sys: 488 ms, total: 1.02 s
Wall time: 791 ms
~~~
{: .output}

The timing that we have recorded for this step is now closer to the one recorded for the serial calculation (the
parallel calculation actually took slightly longer).

> ## Serial vs parallel
>

> > Can you speculate why the parallel calculation actually took longer than its serial version?

Reviewer: Could this be because of the size of the Dask task graph, that it's too large so the overhead of computing the task graph is more than the gains from parallelizing?

I'm not sure if we should make this a challenge, it might be too non-intuitive to those who haven't encountered Dask before. it's already non-intuitive for me!

Another cause could be that dask isn't parallelizing the writing step at all?

Author: Indeed the task graph is relatively large, but also the chunk size that I have used here are small (4 and 32 MB), so the overhead is indeed larger than the gain.  

But you are probably right in this being too non-intuitive for a challenge I will add these arguments just as regular teaching material.

---
title: "Parallel raster computations using Dask"
teaching: 40
exercises: 20
questions:

> - "How can I parallelize computations on rasters with Dask?"

Reviewer 2: ```suggestion
- "How can I run several computations on rasters simultaneously (parallelization) with Dask?"
```

Author: Uhm, this is actually slightly different (we show how to parallelize a single raster calculation, not how to run multiple calculations in parallel). What about:

"How can I exploit parallelization in my raster calculations?"

Reviewer 2: yes, much better, thanks.

- "How can I determine if parallelization improves calculation speed?"

> - "How can I control how to parallelize calculations?"

Reviewer 2: It is not clear to me what is meant by the last question. Can we rephrase it to something like: `What are the good practices in applying a parallelization approach?`

Author: Good point, but maybe that's too general? How about:

"What are good practices in applying parallelization to my raster calculations?"

objectives:

> - "Profile the timing of your raster calculations."

Reviewer 2: ```suggestion
- "Profile the timing of the raster calculations."
```

- "Open raster data as a chunked array."
- "Recognize good practices in selecting proper chunk sizes."
- "Setup raster calculations that take advantage of parallelization."
keypoints:
- "The `%%time` Jupyter magic command can be used to profile calculations."
- "Data 'chunks' are the unit of parallelization in raster calculations."
- "(`rio`)`xarray` can open raster files as chunked arrays."
- "The chunk shape and size can significantly affect the calculation performance."
- "Cloud-optimized GeoTIFFs have an internal structure that enables performant parallel read."
---

> 

Reviewer 2: ```suggestion

# Introduction
```

Very often raster computations involve applying the same operation to different pieces of data. Think, for instance, to
the "pixel"-wise sum of two raster datasets, where the same sum operation is applied to all the matching grid-cells of
the two rasters. This class of tasks can benefit from chunking the input raster(s) into smaller pieces: operations on

> different data blocks can be run in parallel using multiple computing units (e.g., multi-core CPUs), thus potentially

Reviewer 2: ```suggestion
different pieces can be run in parallel using multiple computing units (e.g., multi-core CPUs), thus potentially
```

> speeding up calculations. In addition, working on chunked data can also lead to smaller memory footprints, since one

Reviewer 2: ```suggestion
speeding up calculations. In addition, working on chunked data can lead to smaller memory footprints, since one
```

may bypass the need to store the full dataset in memory by processing it chunk by chunk.

In this episode, we will introduce the use of Dask in the context of raster calculations. Dask is a Python library for

> parallel and distributed computing that provides a framework to work with different data structures, including chunked

Reviewer 2: ```suggestion
parallel and distributed computing and provides a framework to work with different data structures, including chunked
```

Author: Thanks for making me realise this sentence is probably too long - I will actually split it in two!

arrays (Dask Arrays). Dask is well integrated with (`rio`)`xarray` objects, which can use Dask arrays as underlying
data structures.

> ## More Resources on Dask
>
> * [Dask Array](https://docs.dask.org/en/stable/array.html).
> * [Xarray with Dask](https://xarray.pydata.org/en/stable/user-guide/dask.html).
>
{: .callout}

It is important to realize, however, that many details determine the extent to which using Dask's chunked arrays instead
of regular Numpy arrays leads to faster calculations (and lower memory requirements). The actual operations to carry
out, the size of the dataset, and parameters such as the chunks' shape and size, all affects the performance of our
computations. Depending on the specifics of the calculations, serial calculations might actually turn out to be faster!
Being able to time profile your calculations is thus essential, and we will see how to do that in a Jupyter environment
in the next section.

> # Time profiling calculations in Jupyter

Reviewer 2: ```suggestion
# Time profiling in Jupyter
```

> >
> > ~~~
> > band = rioxarray.open_rasterio(band_url, chunks="auto")
> > ~~~
> > {: .language-python}
> >
> > which leads to `(1, 8192, 8192)` chunks (128 MB).
> {: .solution}
{: .challenge}

## Parallel computations

Operations performed on a `DataArray` that has been opened as a chunked Dask array are executed using Dask. Dask
coordinates how the operations should be executed on the individual chunks of data, and runs these tasks in parallel as
much as possible. By default, Dask parallelizes operations on the CPUs that are available on the same machine, but it
can be configured to dispatch tasks on large compute clusters.

Let's now repeat the raster calculations that we have carried out in the previous section, but running calculations in
parallel over a multi-core CPU. We first open the relevant rasters as chunked arrays:

~~~
scl = rioxarray.open_rasterio(scl_href, lock=False, chunks=(1, 2048, 2048))
visual = rioxarray.open_rasterio(visual_href, overview_level=0, lock=False, chunks=(3, 2048, 2048))
~~~
{: .language-python}

Setting `lock=False` tells `rioxarray` that the individual data chunks can be loaded simultaneously from the source by

> the Dask workers. We trigger the download of the data using the `.persist()` method. This makes sure that the downloaded

Reviewer 2: ```suggestion
the Dask workers. 

As the next step, we trigger the download of the data using the `.persist()` method, see below. This makes sure that the downloaded
```

visual = visual.persist(scheduler="threads", num_workers=4)
~~~
{: .language-python}

~~~
CPU times: user 1.18 s, sys: 806 ms, total: 1.99 s
Wall time: 12.6 s
~~~
{: .output}

So downloading chunks of data using 4 workers gave a speed-up of almost 4 times (40.5 s vs 12.6 s)!

Let's now continue to the second step of the calculation. Note how the same syntax as for its serial version is employed
for creating and applying the cloud mask. Only the raster saving includes additional arguments:
* `tiled=True`: write raster as a chunked GeoTIFF.
* `lock=threading.Lock()`: the threads which are splitting the workload must "synchronise" when writing to the same file
  (they might otherwise overwrite each other's output).
* `compute=False`: do not immediately run the calculation, more on this later.

~~~
from threading import Lock
~~~
{: .language-python}

~~~
%%time
mask = scl.squeeze().isin([8, 9])
visual_masked = visual.where(~mask, other=0)
visual_store = visual_masked.rio.to_raster("band_masked.tif", tiled=True, lock=threading.Lock(), compute=False)
~~~
{: .language-python}

~~~
CPU times: user 13.3 ms, sys: 4.98 ms, total: 18.3 ms
Wall time: 17.8 ms
~~~

> {. output}

Reviewer 2: ```suggestion
{: .output}
```

Author: Amazing you caught this! :D
